File: step1/src/knapsack.js
/**
 * Implements the memory function method for the knapsack problem
 * @param {number} capacity - the total amount of weight you can carry
 * @param {Array.<number>} weights - non-negative array of the weights of all the items
 * @param {Array.<number>} values - an array of the values of all the items
 * @returns {number} maximum value that you will be able to carry
 */
function knapsack(capacity, weights, values) {
    if (weights.length !== values.length) {
        throw new Error("The number of weights and values must be equal");
    }

    const n = weights.length; // number of items

    // Create a 2D array to store the maximum value at each capacity and for each item
    const K = new Array(n + 1);
    for (let i = 0; i <= n; i++) {
        K[i] = new Array(capacity + 1).fill(0);
    }

    // Fill the K table. Bottom-up approach
    for (let i = 0; i <= n; i++) {
        for (let w = 0; w <= capacity; w++) {
            if (i === 0 || w === 0) {
                // if no items or capacity is 0, then maximum value is 0
                K[i][w] = 0;
            } else if (weights[i - 1] <= w) {
                // either include it or exclude the current item if its weight is less than or equal to the remaining capacity
                K[i][w] = Math.max(
                    values[i - 1] + K[i - 1][w - weights[i - 1]], // include current item
                    K[i - 1][w] // exclude current item
                );
            } else {
                // If the current item's weight is more than the remaining capacity, exclude it
                K[i][w] = K[i - 1][w];
            }
        }
    }

    return K[n][capacity];
}

module.exports = knapsack;

File: step1/src/Knapsack.java
public class Knapsack {
    public static void main(String[] args) {
        System.out.println(knapsack(new int[] {3, 6, 8},new int[] {50, 60, 100}, 10));
        System.out.println(knapsack(new int[] {3, 6, 8},new int[] {50, 60, 100}, 7));
        System.out.println(knapsack(new int[] {3, 6, 8},new int[] {50, 60, 100}, 0));
        System.out.println(knapsack(new int[] {1, 2, 4, 5 ,7, 8},new int[] {2, 5, 6,10, 13, 16}, 8));
    }

    static int knapsack(int[] weights, int[] values, int maxWeightCarry) {
        int n = weights.length;
        // dimension error
        if(weights.length != values.length){
            return -1;
        }
        // size 0
        if (n == 0 || maxWeightCarry <= 0) {
            return 0;
        }


        // empty knapsack
        int[][] table = new int[n + 1][maxWeightCarry + 1];
        for (int j = 0; j <= maxWeightCarry; j++) {
            table[0][j] = 0;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= maxWeightCarry; j++) {
                if (weights[i - 1] > j) {
                    table[i][j] = table[i - 1][j];
                } else {
                    int currentValue = table[i - 1][j];
                    int candidateValue = table[i - 1][j - weights[i - 1]] + values[i - 1];
                    if (candidateValue > currentValue) {
                        table[i][j] = candidateValue;
                    } else {
                        table[i][j] = currentValue;
                    }
                }
            }
        }

        // show table
        for (int i = 0; i < table.length; i++) {
            for (int j = 0; j < table[i].length; j++) {
                System.out.print(table[i][j] + " ");
            }
            System.out.println();
        }

        return table[n][maxWeightCarry];
    }

}
